### PR TITLE
Add fp8 ddp and remove ready_to_all_reduce

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Add custom ddp which supports fp8 and compute/communication overlap

**Major Revision**
- Add custom ddp which supports scaling tensor and compute/communication overlap
- Add UT for custom ddp
- Remove ready_to_all_reduce from model state

Test it use mnist_ddp:
```bash
torchrun --nproc_per_node=8 mnist_ddp.py --enable-msamp --opt-level=O2
```

### CODE REVIEW

File: msamp/nn/linear.py [hunk 1]
```diff
@@ -176,7 +176,7 @@ def replace(cls, model, weight_qtype=Dtypes.kfloat16, src_rank=0, group=None):
         model = cls._replace(model, weight_qtype)
         fp8_named_weights = [(k, p) for k, p in model.named_parameters() if isinstance(p, ScalingParameter)]
 
-        fp8_names = [k for k, _ in fp8_named_weights]
+        # fp8_names = [k for k, _ in fp8_named_weights]
```
Comment (wkcn) [inline msamp/nn/linear.py:179]: remove this comment

File: examples/mnist_ddp.py [hunk 2]
```diff
@@ -191,6 +191,7 @@ def main():
     if args.enable_msamp:
         print(f'msamp is enabled, opt_level: {args.opt_level}')
         model, optimizer = msamp.initialize(model, optimizer, opt_level=args.opt_level)
+        msamp.nn.model_state.use_fp8_ddp = True
```
Comment (wkcn) [inline examples/mnist_ddp.py:194]: Please set `msamp.nn.model_state.use_fp8_ddp = True` automatically when calling `msamp.nn.distributed.FP8DistributedDataParallel`.

File: msamp/nn/linear.py (hunk 1 above)
Comment (tocean) [reply msamp/nn/linear.py]: Done.

File: examples/mnist_ddp.py (hunk 2 above)
Comment (tocean) [reply examples/mnist_ddp.py]: Good idea. Fixed.

File: msamp/nn/__init__.py [hunk 3]
```diff
@@ -9,6 +9,8 @@
 from msamp.nn.linear import FP8Linear, LinearReplacer
 from msamp.nn import functional
 from msamp.nn.clip_grad import clip_grad_norm_
+from msamp.nn import distributed
 del functional
+del distributed
```
Comment (abuccts) [inline msamp/nn/__init__.py:14]: why this is needed?

File: msamp/nn/distributed.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+                buf.copy_(fp8_grad)
+                params[i].grad = ScalingTensor(buf, meta)
+                params[i].grad.div_(world_size)
+                bucket_offset += grad_numel
+
+            flat_fp8_grads = self.buffer.narrow(0, bucket_start, bucket_end - bucket_start)
+
+            # step 5: allreduce the gradients
+            torch.cuda.default_stream().wait_stream(self.reduction_stream)
+            FP8Op.enable_fp8(wgrad_qtype)
```
Comment (abuccts) [inline msamp/nn/distributed.py:186]: do you want to rename FP8Op to distributed related names like distOp?

**Review by abuccts — COMMENTED** [review]
Will this PR replace https://github.com/Azure/MS-AMP/pull/75?

File: msamp/nn/distributed.py (hunk 4 above)
Comment (tocean) [reply msamp/nn/distributed.py:186]: Yes, it is in another PR(https://github.com/Azure/MS-AMP/pull/94). Will replace it with DistOp in that PR after this PR merged.

File: msamp/nn/__init__.py (hunk 3 above)
Comment (tocean) [reply msamp/nn/__init__.py:14]: `from msamp.nn import distributed` will call `torch.nn.parallel.DistributedDataParallel = FP8DistributedDataParallel`, which is neccessary. However, just like msamp.nn.functional, msamp.nn.distributed is something internal in MS-AMP and should not be exposed. So just as what we did to functional, we `del distributed` after importing it.

**Comment by tocean** [comment]
> Will this PR replace #75?

Yes, already synced with kan and will close #75 after this PR merged.

File: msamp/nn/__init__.py [hunk 5]
```diff
@@ -9,6 +9,8 @@
 from msamp.nn.linear import FP8Linear, LinearReplacer
 from msamp.nn import functional
 from msamp.nn.clip_grad import clip_grad_norm_
+from msamp.nn import distributed
```
Comment (guoshzhao) [inline msamp/nn/__init__.py:12]: Can we put attributes from same module together?
`from msamp.nn import functional, distributed`

Comment (tocean) [reply msamp/nn/__init__.py]: Sure. Fixed.

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**